v2/pkg/pool/limitbufferpool.go

```diff
@@ -0,0 +1,98 @@
+package pool
+
+import (
+	"bytes"
+	"context"
+	"time"
+)
+
+// LimitBufferPool is a pool of buffers that is limited in size and is limiting the max size of buffers that should be recycled
+// In addition, it runs a GC runtime that randomly resets a buffer every second to keep the memory usage low when usage is low
+// This is an alternative to sync.Pool, which can grow unbounded rather quickly
+type LimitBufferPool struct {
+	buffers []*ResolvableBuffer
+	index   chan int
+	options LimitBufferPoolOptions
+}
+
+type ResolvableBuffer struct {
+	Buf *bytes.Buffer
+	idx int
+}
+
+type LimitBufferPoolOptions struct {
+	// MaxBuffers limits the total amount of buffers that can be allocated for printing the response
+	// It's recommended to set this to the number of CPU cores available, or a multiple of it
+	// If set to 0, the number of CPU cores is used
+	MaxBuffers int
+	// MaxBufferSize limits the size of the buffer that can be recycled back into the pool
+	// If set to 0, a limit of 10MB is applied
+	// If the buffer cap exceeds this limit, a new buffer with the default size is created
+	MaxBufferSize int
+	// DefaultBufferSize is used to initialize the buffer with a default size
+	// If set to 0, a default size of 8KB is used
+	DefaultBufferSize int
+	// GCTime is the time interval to run the GC to reset the buffer
+	GCTime time.Duration
+}
+
+func NewLimitBufferPool(ctx context.Context, options LimitBufferPoolOptions) *LimitBufferPool {
+	if options.MaxBufferSize == 0 {
+		options.MaxBufferSize = 1024 * 1024 * 10 // 10MB
+	}
+	if options.DefaultBufferSize == 0 {
+		options.DefaultBufferSize = 1024 * 8 // 8KB
+	}
+	if options.MaxBuffers == 0 {
+		options.MaxBuffers = 8
+	}
+	if options.GCTime == 0 {
+		options.GCTime = time.Second
+	}
+	pool := &LimitBufferPool{
+		buffers: make([]*ResolvableBuffer, options.MaxBuffers),
+		index:   make(chan int, options.MaxBuffers),
+		options: options,
+	}
+	for i := range pool.buffers {
+		pool.buffers[i] = &ResolvableBuffer{
+			idx: i,
+		}
+		pool.index <- i
+	}
+	go pool.runGC(ctx)
+	return pool
+}
+
+func (p *LimitBufferPool) runGC(ctx context.Context) {
+	ticker := time.NewTicker(p.options.GCTime)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			b := p.Get()
```

Comment:
This means, it will reset a buffer every second. You can't ensure that you get an existing buffer. In the worst you allocate and reset it in one step. I have never seen this in real world. How can we be sure that this works out in advanced use case?